### PR TITLE
[Testing:InstructorUI] Add cypress test for Rainbow Grading

### LINKED ISSUE #10042: Cypress Test for RainbowGrades
**What problem are you trying to solve with Submitty**

Currently in submitty RainbowGrades is not tested. 
Hence, we cannot spot any issue if caused by updated code (mostly in rainbowgrades repo)

**Describe the way you'd like to solve this problem**

We should add a RainbowGrades cypress test.
It does not have to be a complete model for now.
Start from baby step 
For example, check if aphacker has correct grade for closed_homework after running generate grade summaries. 
And, check if the gradebook and individual grade values matches correctly. 
These are some general ideas, dev can add much more reasonable testing.

### PR DESCRIPTION
**Pull request by charvimehradu**
### Please check if the PR fulfills these requirements:

* [x] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant
* [x] Screenshots are attached to Github PR if visual/UI changes were made

### What is the current behavior?
Closes #10042 

Currently, there are no tests written for testing the Rainbow Grades in Submitty. Hence, we cannot spot any issue if caused by updated code.

### What is the new behavior?
Added cypress test to:

- Test that for sample- rainbow grades can be configured automatically.
- Test that for sample- rainbow grades can generate grade summaries and customize what to show.
- Check if the results/grades are some for both the instructor and student UI.

https://github.com/Submitty/Submitty/assets/121369234/ffeaa3fa-64a7-4a4b-a3a6-8d8b71bed494

### Other information?
The configuration for rainbow grades get changed after running the test. It will be a good practice to revert those changes back to the default.

### CODE REVIEW

**Comment by manzil-infinity180** [comment]
@charvimehradu try to use less repetitive code and one suggestion if you are raising a pr for test if you attach a screenshot of completion of test or video .. may these help a lot for maintainer/mentor/admin

**Comment by charvimehradu** [comment]
Hi @manzil-infinity180, thank you for your suggestions. I have refactored the code to reduce repetition.

**Comment by williamjallen** [comment]
This appears to be a duplicate of https://github.com/Submitty/Submitty/pull/10181, and likely suffers from the same blocker.  Rainbow Grades is not installed in our CI system, which means that this test will work locally, but will fail in our CI.  Given how critical Rainbow Grades is, it would probably be beneficial to investigate getting Rainbow Grades set up in our CI.

**Comment by charvimehradu** [comment]
Hi @williamjallen. Indeed the credits for the actual PR goes to @VVVITH.

Do you think it would be possible for me to help with setting it up in the CI? Any guidance or leads on how I can contribute to resolving this would be greatly appreciated!

**Comment by williamjallen** [comment]
> Do you think it would be possible for me to help with setting it up in the CI? Any guidance or leads on how I can contribute to resolving this would be greatly appreciated!

Getting Rainbow Grades (and potentially some of the other child repos like Lichen and AnalysisTools) working in the CI would be a great improvement.  I'm not entirely sure what it would require at this point, but you're more than welcome to work on it.  If you need help or want to discuss it further, feel free to post on Zulip or create an issue for this.

File: site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js
Comment (ziesski) [inline site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js:39]: Hard-coded waits are bad practice for  test automation because they can lead to inefficiency and brittleness in tests. Instead, you can utilize Cypress commands to handle them. 

For example 
` cy.get('#grade-summaries-button').click();
   cy.get('#grade-summary').should('be.visible');
`

File: site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js
Comment (ziesski) [inline site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js:51]: same thing applies here

File: site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js
Comment (ziesski) [inline site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js:60]: same thing applies here

**Review by ziesski — DISMISSED** [review]

File: site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js
Comment (charvimehradu) [reply site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js]: Thanks for the suggestion, I have implemented it.

**Comment by williamjallen** [comment]
@charvimehradu It looks like your new test is failing.

**Comment by manzil-infinity180** [comment]
@williamjallen can you elaborate why this happen ?

**Comment by williamjallen** [comment]
> @williamjallen can you elaborate why this happen ?

I'm not quite sure what happened.  The forked repository is out of date, and something about one of the merge commits caused GitHub to display things weirdly.  Merging with our main branch seems to have resolved the problem.

File: site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js
Comment (IDzyre) [inline site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js:25]: Is it possible to do it without any explicit timeout?

File: site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js
Comment (IDzyre) [inline site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js:15]:
```suggestion
        cy.get('[data-testid="display-grade-summary"]').then(($checkbox) => {
```

File: site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js
Comment (IDzyre) [inline site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js:12]: Is this all necessary? Are you expecting the state of the VM to be different for each test?

File: site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js
Comment (IDzyre) [inline site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js:34]: remove console log

**Review by IDzyre — CHANGES_REQUESTED** [review]

File: site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js
Comment (charvimehradu) [reply site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js:25]: We can not move to the next segment of the tests without generating the summary and it does take some time. I am not sure if there exists another way to generate the rainbow grades.

File: site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js
Comment (IDzyre) [inline site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js:15]:
```suggestion
        cy.get('[data-testid="display-grade_summary"]').then(($checkbox) => {
```

This is one of the exception cases, as this is a procedurally generated testid, using the display value, that uses underscores.

**Review by IDzyre — CHANGES_REQUESTED** [review]

**Comment by manzil-infinity180** [comment]
@IDzyre i think this pr messed up something some unusal commit has added here i don't have any idea of it

**Comment by IDzyre** [comment]
> @IDzyre i think this pr messed up something some unusual commit /file has added here i don't have any idea of it

This should be fixed now.

**Review by IDzyre — CHANGES_REQUESTED** [review]

**Comment by manzil-infinity180** [comment]
Please the see the CI logs for commit
1) removed checked (Test failed for line 10) it means the checkbox is enabled default
and then 
2) added for checkbox - test failed for line 8 

how ??
Is any way to see the cypress images that are failing in CI pipeline???

**Comment by IDzyre** [comment]
> Please the see the CI logs for commit
> 
> 1. `removed checked` (Test failed for line 10) it means the checkbox is enabled default
>    and then
> 2. `added for checkbox` - test failed for line 8
> 
> how ??
> 
> Is any way to see the cypress failed screenshot that are failing in CI pipeline???

Go to the Summary of the CI, scroll down, and there are cypress-screenshots and cypress-videos in the Artifacts section

**Comment by IDzyre** [comment]
Also, have you checked to see if the CI gradeable has submissions you can test?

**Comment by IDzyre** [comment]
@manzil-infinity180 could you add the items tested in https://github.com/Submitty/Submitty/pull/10181 to this PR, so there is only one PR for rainbow grades tests that has a more complete test?

**Comment by manzil-infinity180** [comment]
> @manzil-infinity180 could you add the items tested in #10181 to this PR, so there is only one PR for rainbow grades tests that has a more complete test?

Ok i will add more items ( from https://github.com/Submitty/Submitty/pull/10181)

File: site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js
Comment (IDzyre) [inline site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js:60]: Would it be more consistent with this file to use a .forEach() statement on an array instead of a bunch of .and() statements? Not that it is bad practice, just wondering for consistency.

**Review by IDzyre — COMMENTED** [review]
One comment about consistency, but looks good to merge, and I will approve once it is addressed.

**Comment by IDzyre** [comment]
Once the merge conflicts are addressed I will approve and merge

File: site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js
Comment (IDzyre) [inline site/cypress/e2e/Cypress-Gradeable/rainbow_grading.spec.js:1]:
```suggestion
import { skipOn } from '@cypress/skip-test';

skipOn(Cypress.env('run_area') === 'CI', () => {
    describe('Test Rainbow Grading', () => {
```

Cypress tests are failing I believe due to the lack of submissions in the CI, so to get this merged before they are added, add this to skip on CI, and we will fix the tests if they break when we add submissions.

**Review by IDzyre — CHANGES_REQUESTED** [review]

**Review by IDzyre — APPROVED** [review]
Approving and merging now, we can revisit this test at a later date once we get submissions in CI.

**Pull request merged**
